Blank Case Search

esoergel opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/USH-1150
https://dimagi-dev.atlassian.net/browse/USH-1105
https://docs.google.com/document/d/1sLHXEFPfj56zcLw8ZZPDXMVOUxqHiIpxKIcp2hE2ueE/edit#
To be deployed with https://github.com/dimagi/formplayer/pull/987

I'm sorry to say it might be easiest to review as a whole - there was a decent amount of churn to make sense of the JS.  I'm still familiarizing myself with this codebase, so feel free to provide opinionated feedback.

See product description below.

## Feature Flag
"Enable synchronous mobile searching and case claiming"

## Product Description
This adds the ability to search for cases missing specific values.  Here's the configuration screen:

![image](https://user-images.githubusercontent.com/2367539/128071032-a274483a-aa51-424d-a34b-c381249ab939.png)

And here's how it looks in the UI:

![image](https://user-images.githubusercontent.com/2367539/131135976-01dbf809-3261-4510-8287-1e6a4d33f7c3.png)

Values are considered to be blank/empty if they are not set on the case or are set to the empty string.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3205

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

esoergel committed:
Only display help text when relevant

The check box is hidden when appearanceFinal is not fixture, and the
message doesn't make sense without that option

esoergel committed:
Make "blank search" checkbox and backend

esoergel committed:
Pass config through in suite.xml

esoergel committed:
DRY up booleans

esoergel committed:
Add ability to search for blank values to webapps

esoergel committed:
indent

esoergel committed:
Fix table - keep it from resizing

esoergel committed:
Add (empty) option to select questions

esoergel committed:
Add blank search for date fields

esoergel committed:
Disable input field when searching for blank

esoergel committed:
disable blank search option for geocoder fields

esoergel committed:
Provide explicit handling of blank select values

esoergel committed:
Correctly identify empty multiselects

esoergel committed:
Show checkbox for all field types

esoergel committed:
The checkbox should not disable the input field

esoergel committed:
Make checkbox function properly

Ie, not get wiped when parts of the UI are redrawn, submitting the correct
values to formplayer, etc.

esoergel committed:
Do daterange queries with the rest of 'em

esoergel committed:
Extract fuzzy properties queries

esoergel committed:
Search fuzzy properties with multiple case types

This test fails without the accompanying code change. Because the second
case type doesn't have fuzzy properties defined, 'fuzzy' was getting
reset.  If the case types were passed in in the reverse order, it
would've succeeded.

esoergel committed:
extract code that cleans the strings

esoergel committed:
remove date query method

esoergel committed:
Extract case search logic

Note that this generates a differently structured query, but it should
behave the same

esoergel committed:
Replace duplicate implementations with standard

esoergel committed:
Pull out query construction from mutation

esoergel committed:
Support blank search as one of multiple values passed

This is to support the "include blank values" checkbox in webapps, which will
pass in the entered value and an empty string, to be OR'd together

esoergel committed:
Move multi param check to after blanks are rm'd

esoergel committed:
Don't include '' for geocoder either

esoergel committed:
Make geocoder questions propagate to their members

esoergel committed:
Move behind USH flag

esoergel committed:
Update checkbox description

esoergel committed:
Rejigger checkbox handling for sticky search

esoergel committed:
Make input fields update the model directly

Shout out to @orangejenny for talking through the proper approach and
for finding out how to access the parent view

esoergel committed:
ensure searchForBlank is always set on the model

esoergel committed:
Allow blank search for related case properties

esoergel committed:
call changeDropdown from inside changeQueryField

They were fighting over who has control of that model

esoergel committed:
Clear checkboxes as well

esoergel committed:
Move geocoder init to parent view

It was currently called for each field, even though it iterates over all
fields.  I also extracted out the helper functions which don't need `self`

esoergel committed:
Reinitialize geocoder on relevant model changes

esoergel committed:
Restore user input after redraw

esoergel committed:
Split out mapbox widget initialization

esoergel committed:
Only initialize once

Otherwise if there's an init without a redraw (like on clear, when
there's nothing populated), it'll make two widgets

orangejenny commented:
FYI I'm about halfway through reviewing this and should be able to finish it tomorrow.

esoergel committed:
Sticky checkboxes also for fields with defaults

orangejenny commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -13,29 +13,41 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             }
+            var values = value.split(selectDelimiter),
+                searchForBlank = _.contains(values, ""),
+                values = _.filter(values, function (val) { return val !== ""; });
```
`_.without(values, "")` would be nice & concise here

orangejenny commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -329,15 +316,30 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
 
+        toggleBlankSearch: function (e) {
+            var self = this,
+                searchForBlank = $(e.target).prop('checked'),
```
Nit: I'm a proponent of always using `currentTarget` - doesn't matter here, but `target` can be a descendant of the element that has the event handler.

esoergel replied:
thanks, I wasn't aware

orangejenny reviewed (commented):
This looks great.

esoergel commented on corehq/apps/app_manager/static/app_manager/js/details/case_claim.js:
```diff
@@ -259,6 +263,7 @@ hqDefine("app_manager/js/details/case_claim", function () {
 
         if (searchProperties.length > 0) {
             for (var i = 0; i < searchProperties.length; i++) {
+                // searchProperties is a list of CaseSearchProperty objects
```
Oh yeah, that boilerplate was a bit annoying.  I also had trouble figuring out where the raw data from formplayer was initialized into a JS object - would be nice to have that happen in just one place, but it's kinda scattered across a few of these.

esoergel committed:
Simplify html

esoergel commented on corehq/apps/case_search/utils.py:
```diff
@@ -119,27 +123,30 @@ def _add_case_property_queries(self):
+    def _patterns_to_remove(self):
+        patterns_by_property = defaultdict(list)
+        for pattern in self.config.ignore_patterns.filter(domain=self.domain, case_type__in=self.case_types):
+            patterns_by_property[pattern.case_property].append(re.escape(pattern.regex))
```
huh, yeah I think you're right - does this mean these patterns don't actually work as regexps?

orangejenny replied:
I believe so. The settings page where you configure these expressions literally uses `+`, an invalid regex, as an example.

esoergel replied:
That's...unfortunate.  It does look like it only does exact string matches, and that's clear from what's presented to the user, at least.

esoergel commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html:
```diff
@@ -92,6 +92,8 @@ <h4 class="modal-title">
                       </div>
                   </div>
+                {% endif %}
+                {% if js_options.default_value_expression_enabled %}
```
This is the way the other things under the USH_CASE_CLAIM_UPDATES flag are gated, so I just went along with it.  It also checks a minimum app version, though I guess that's not really relevant for webapps.

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -358,11 +358,10 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
-                $(this).trigger('change.select2');
+            var self = this;
+            _.each(self.collection.models, function (model) {
+                model.set('value', '');
```
yeah, kinda indirectly - the model change triggers a re-render, so `onRender` gets called, which initializes select2.  Now that I'm writing that out, I wonder if I could've used that same logic to call `initMapboxWidget` when needed.

esoergel committed:
PR feedback

esoergel committed:
Update corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

esoergel committed:
Make geocoder widget also work when using mouse

There seems to be some kind of timing issue between when geocoder
returns a response and when changeQueryField fires.  By moving the model
change into the callback, we can be sure it happens only afterwards.

esoergel commented:
@orangejenny pinging for a review of the last couple commits - should be tiny, but this now passed QA and is ready for merge

snopoke commented on corehq/apps/es/tests/test_case_search_es.py:
```diff
@@ -557,6 +557,25 @@ def _assert_related_case_ids(self, cases, paths, ids):
+            None,
+            ['c1', 'c2']
+        )
+        fuzzy_properties.delete()
+        config.delete()
```
These should use `self.addCleanup` and be above the assert.

snopoke reviewed (commented):
Same comment for other tests.

snopoke commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -13,87 +13,57 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
+        geocoderItemCallback = function (addressTopic, model) {
             return function (item) {
+                model.set('value', item.place_name);
+                initMapboxWidget(model);
```
@esoergel do you recall why it was necessary to call `initMapboxWidget`?

This function handles events from the widget so I don't see how calling `initMapboxWidget` make sense.

esoergel replied:
I don't recall, but I agree it looks pretty weird.  It was tricky to manage control flow - one guess is maybe something cleared out the input in the widget and re-initializing the widget set it back?
